azure/scope/managedmachinepool.go

```diff
@@ -134,11 +134,26 @@ func (s *ManagedMachinePoolScope) Name() string {
 	return s.InfraMachinePool.Name
 }
 
+// SetSubnetName updates AzureManagedMachinePool.SubnetName if AzureManagedMachinePool.SubnetName is empty with s.ControlPlane.Spec.VirtualNetwork.Subnet.Name.
+func (s *ManagedMachinePoolScope) SetSubnetName() {
+	s.InfraMachinePool.Spec.SubnetName = getAgentPoolSubnet(s.ControlPlane, s.InfraMachinePool)
```

Comment:
`getAgentPoolSubnet` is just setting it to itself if `s.InfraMachinePool.Spec.SubnetName` is already set. It might be more readable to do:

```
func (s *ManagedMachinePoolScope) SetSubnetName() {
        if s.InfraMachinePool.Spec.SubnetName == nil {
	    s.InfraMachinePool.Spec.SubnetName = pointer.StringPtr(controlPlane.Spec.VirtualNetwork.Subnet.Name)
        }
}
```

And then always use `s.InfraMachinePool.Spec.SubnetName` as the subnet name value for the agent pool (get rid of the `getAgentPoolSubnet` func entirely).

Reply:
We cannot get rid of that function as it is used in buildAgentPoolSpec. The AMCP reconciler kicks in first and AMMP will not be triggered until AMCP is initialized. So while creating the AMCP we need that value.

Reply:
updated the functionality for better readability

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/79c47714bfd637d9d979494af08310e712c855cf/controllers/azuremanagedmachinepool_controller.go#L191

azuremanagedmachinepool will not even start reconciling until azuremanagedcontrolplane is initialized and amcp will not be initialized until it has a subnetName it would be a deadlock again.

Reply:
Ahh I didn't realize that `buildAgentPoolSpec` was also called by managedcontrolplane.go... In that case it would be a problem since the subnet name might not be set yet. This is a bit unusual and due to the fact that we can't create an AKS cluster with no node pools (I was just talking about this with @matthchr and @nojnhuh the other day) which causes us to not have a clean separation of concern between controllers and CRDs (the AzureManagedControlPlane controller reconciles the AzureManagedMachinePool objects too, on the first reconcile).

Reply:
Not sure if there is a better/more elegant way to solve this for now, besides doing the same defaulting from the managedcontrolplane reconciler too, or switching back to the webhook approach to ensure that the SubnetName is always set on the AzureManagedMachinePool before the controllers run.

I know previously you ran into an issue with the webhook defaulting where order of creation is not always guaranteed when the objects are all created together, one potential fix for that is to retry getting the AzureManagedControlPlane a few times like https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_default.go#L187 is doing.

Reply:
What is the issue with the current implementation ? Why do you think it is not good w.r.t code quality ? Why are we looking for a better approach ? Just trying to understand what is missing as we have been moving back and forth on this.

Reply:
I am open to move the logic to the webhook in this PR. The idea is to add a wait in AMMP webhook until AMCP is created correct ?

Reply:
Let's merge this as-is and explore moving to a webhook in a follow-up PR.

Reply:
ack